### PR TITLE
#7521 - Introduced upper batch limit and batching for person queries

### LINKED ISSUE #7521: Pull data to the mobile app in batches [5]
### Feature Description

Functional data (such as cases, contacts, persons, samples, etc.) is currently synchronised in one single request.

### Problem Description

Especially countries that use a single central SORMAS server run into significant problems here - not only during the initial synchronisation, but potentially also during other syncs, because there's too much functional data. Trying to pull this data from the server with low bandwidth can still lead to timeouts.

### Proposed Change (WIP)

Synchronise all production data in batches.

#### Calculate the approximate size of every entity
* For each entity that is synchronized this way (see `SychronizeDataAsync#synchronizeChangedData`, excluding outbreaks, disease configurations and customizable enum values), create an example entity on your local server (or a test server) with as much data filled in as possible; if there are sublists that are synchronized alongside the main entity (e.g. contact details for persons, exposures for cases or contacts), create a few entries in these lists as well.
* Do a ReST call that only retrieves this example entity (via Postman or in the browser); copy the returned JSON into a text file and note down the size of the text file in bytes.
* For each of these entities, add a final String to the DTO class that is called `APPROXIMATE_JSON_SIZE`, and set it to the size that you noted down for the respective entity.

#### Calculate the gzip compression factor
* Create ten cases on your local server (or a test server); the detail level of these cases doesn't matter, so you can use the dev mode data generator
* Do a ReST call that only retrieves these ten cases; copy the returned JSON into a text file and note down the size of the text file.
* Compress the text file using gzip with compression level 1; note down the size of the resulting compressed file.
* Calculate the compression factor by comparing the two file sizes, and store it in a final String called `JSON_COMPRESSION_FACTOR` in `SynchronizeDataAsync`.

#### Prepare the Android backend
* Change the synchronization methods to pull data for each individual entity until there's no more data left to be pulled
  * This can be determined by checking whether the number of returned entities < the requested number of entities
* When inserting pulled data, keep the change date of the last request, as well as the change date and UUID of the entity that was inserted last
  * When change date of the last request != change date of the last synchronized entity (i.e. change date of the new request), pull the next batch of data WITHOUT specifying the last synchronized UUID
  * When change date of the last request == change date of the last synchronized entity (i.e. change date of the new request), pull the next batch of data INCLUDING the last synchronized UUID
* Reduce the synchronization timeout to 5 minutes

#### Determine the size of the requested batch
To determine the `size` of the individual requests, perform the following steps:

* Request the current available bandwidth in kbps (see https://stackoverflow.com/questions/55159487/android-how-to-determine-network-speed-in-android-programmatically) and multiply it by 60 to see how many kb can approximately be transferred within one minute.
* Determine how many entities can be synchronized within one minute by dividing the available kbpm (kilobits per minute) by the APPROXIMATE_JSON_SIZE of the entity (make sure to convert this to kilobits) multiplied with the JSON_COMPRESSION_FACTOR; always use 10 as the minimum.

### PR DESCRIPTION
**Pull request by MateStrysewske**
Fixes #7521

### CODE REVIEW

File: sormas-app/app/src/main/java/de/symeda/sormas/app/rest/RetroProvider.java [hunk 1]
```diff
@@ -308,7 +309,10 @@ public static Integer getNumberOfEntitiesToBePulledInOneBatch(long approximateJs
 		double compressedJsonSizeInBits = approximateJsonSizeInBytes * 8 / JSON_COMPRESSION_FACTOR;
 		int batchSize =
 			Math.toIntExact(Math.round(getNetworkDownloadSpeedInKbps(context) * ASSUMED_TRANSFER_TIME_IN_SECONDS * 1024 / compressedJsonSizeInBits));
-		return batchSize < 10 ? 10 : batchSize;
+		// Restrict the batch size to a maximum value to avoid backend queries leading to a timeout
+		batchSize = Math.min(MAX_BATCH_SIZE, Math.max(10, batchSize));
+		batchSize = (int) (10 * Math.sqrt(batchSize / 10f));
+		return batchSize;
```
Comment (MartinWahnschaffe) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/rest/RetroProvider.java:315]: Limiting the batch size to 500 has to be done last!

Comment (MateStrysewske) [reply sormas-app/app/src/main/java/de/symeda/sormas/app/rest/RetroProvider.java:315]: True, no idea why I wrote it like that :D

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java [hunk 2]
```diff
@@ -479,7 +479,7 @@ public Predicate buildCriteriaFilter(PersonCriteria personCriteria, PersonQueryC
 				immunizationPersonsQuery.where(immunizationPersonsFilter);
 			}
 			immunizationPersonsQuery.distinct(true);
-			immunizationPersonsResultList = em.createQuery(immunizationPersonsQuery).getResultList();
+			immunizationPersonsResultList = getBatchedQueryResults(cb, immunizationPersonsQuery, personsRoot, batchSize);
 		}
 
 		List<Person> travelEntryPersonsResultList = new ArrayList<>();
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java:485]: please also fix the travel entry batch size comment

**Review by MartinWahnschaffe — APPROVED** [review]

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java [hunk 3]
```diff
@@ -409,7 +409,7 @@ public Predicate buildCriteriaFilter(PersonCriteria personCriteria, PersonQueryC
 			casePersonsQuery.where(casePersonsFilter);
 		}
 		casePersonsQuery.distinct(true);
-		List<Person> casePersonsResultList = em.createQuery(casePersonsQuery).getResultList();
+		List<Person> casePersonsResultList = getBatchedQueryResults(cb, casePersonsQuery, personsRoot, batchSize);
```
Comment (syntakker) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java:412]: `personsRoot` => `casePersonsRoot`

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java [hunk 4]
```diff
@@ -431,7 +431,7 @@ public Predicate buildCriteriaFilter(PersonCriteria personCriteria, PersonQueryC
 			contactPersonsQuery.where(contactPersonsFilter);
 		}
 		contactPersonsQuery.distinct(true);
-		List<Person> contactPersonsResultList = em.createQuery(contactPersonsQuery).getResultList();
+		List<Person> contactPersonsResultList = getBatchedQueryResults(cb, contactPersonsQuery, personsRoot, batchSize);
```
Comment (syntakker) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonService.java:434]: `personsRoot` => `contactPersonsRoot`

**Review by MartinWahnschaffe — APPROVED** [review]

**Pull request merged**